### PR TITLE
code from previos tasks

### PR DESCRIPTION
**Pull request by Mprogrammer2020**

### CODE REVIEW

File: src/components/Chats/Chat/Message.js [hunk 1]
```diff
@@ -4,15 +4,21 @@ import { forwardRef } from 'react';
 import './Message.css';
 
 const Message = forwardRef(
-  ({ contents: { created_at, direction, body, isSending } }, ref) => {
+  ({ contents: { created_at, direction, body, isSending, mms_content } }, ref) => {
+
+    if (mms_content) mms_content = JSON.parse(mms_content)
```
Comment (jeznag) [inline src/components/Chats/Chat/Message.js:9]: can you use

```
export function safelyParseJSON(
  jsonString: string,
  useCase: string,
  fallbackResponseIfError: any,
  logError = false
) {
  try {
    return JSON.parse(jsonString);
  } catch (e) {
    const errorWithMoreInfo = new Error(e);
    errorWithMoreInfo.message = `${e.message}${useCase} ${JSON.stringify(
      jsonString
    )}`;
    if (logError) {
      (window as any).Sentry.captureException(errorWithMoreInfo);
    }

    return fallbackResponseIfError;
  }
}
```

File: src/components/Chats/Chat/Message.js [hunk 2]
```diff
@@ -4,15 +4,21 @@ import { forwardRef } from 'react';
...
+
     return (
       <div
         ref={ref}
         style={{ marginBottom: '10px' }}
         className={`message ${direction === 'outbound' && 'message__sender'}`}>
         <section>
           <div>
-            <p>{body} </p>
+            {mms_content && mms_content.length > 0 ?
```
Comment (jeznag) [inline src/components/Chats/Chat/Message.js:18]: also need to check that it's an array

File: src/components/Chats/Chat/ChatInput.js [hunk 3]
```diff
@@ -85,14 +91,69 @@ const ChatInput = (props) => {
     }
   };
 
+  const sendImages = async () => {
+    inputRef.current.click()
+    // ImagePicker.hasReadPermission(res => {
```
Comment (jeznag) [inline src/components/Chats/Chat/ChatInput.js:96]: should this code still be uncommented?

File: src/components/Chats/Chat/ChatInput.js [hunk 4]
```diff
@@ -85,14 +91,69 @@ const ChatInput = (props) => {
     }
   };
 
+  const sendImages = async () => {
+    inputRef.current.click()
```
Comment (jeznag) [inline src/components/Chats/Chat/ChatInput.js:95]: doesn't seem like a normal react way to handle it?

File: src/components/Chats/Chat/ChatInput.js [hunk 5]
```diff
@@ -85,14 +91,69 @@ const ChatInput = (props) => {
...
+    e
+  ) => {
+    console.log(e.target.files)
+    let inputImages = []
+    for (let i = 0; i < e.target.files.length; i++) {
+      inputImages.push(e.target.files[i])
+    }
+    // uploads is an array that would hold all the post methods for each image to be uploaded, then we'd use axios.all()
+
+    const uploads = inputImages.forEach((image) => {
```
Comment (jeznag) [inline src/components/Chats/Chat/ChatInput.js:117]: this wouldn't wait for all the uploads. Use `await Promise.all(inputImages.map..`

File: src/components/Chats/Chat/ChatInput.js [hunk 6]
```diff
@@ -85,14 +91,69 @@ const ChatInput = (props) => {
...
+    // })
+  }
+
+  const handleUploadImages = (
+    e
+  ) => {
+    console.log(e.target.files)
+    let inputImages = []
+    for (let i = 0; i < e.target.files.length; i++) {
+      inputImages.push(e.target.files[i])
```
Comment (jeznag) [inline src/components/Chats/Chat/ChatInput.js:113]: this seems to be the same as doing `inputImages = e.target.files.slice()`

File: src/components/Chats/Chat/ChatInput.js [hunk 7]
```diff
@@ -85,14 +91,69 @@ const ChatInput = (props) => {
...
+      console.log(image)
+      // our formdata
+      const formData = new FormData();
+      formData.append('file', image);
+      formData.append('tags', 'twilio');
+      formData.append('upload_preset', 'twilio');
+      formData.append('api_key', '173952541553444');
+      formData.append('timestamp', Date.now().toString());
+
+      return fetch('https://api.cloudinary.com/v1_1/lktvgsltz/upload',
```
Comment (jeznag) [inline src/components/Chats/Chat/ChatInput.js:127]: use await

File: src/components/Chats/Chat/ChatInput.js [hunk 8]
```diff
@@ -85,14 +91,69 @@ const ChatInput = (props) => {
...
+      className="ion-align-items-center"
       style={{
         padding: '10px 20px',
         '--border-radius': '20px',
         '--background': '#f5f5f5'
       }}>
-      <IonCol>
+      <IonButton
+        fill="clear"
+        onClick={() => { inputRef.current.click() }}>
```
Comment (jeznag) [inline src/components/Chats/Chat/ChatInput.js:152]: why do you need to use inputRef? You can just call a handler directly

File: src/components/Chats/Chat/Message.js (hunk 1 above)
Comment (Mprogrammer2020) [reply src/components/Chats/Chat/Message.js:9]: ok sure

File: src/components/Chats/Chat/Message.js (hunk 2 above)
Comment (Mprogrammer2020) [reply src/components/Chats/Chat/Message.js:18]: ok sure

File: src/components/Chats/Chat/ChatInput.js (hunk 5 above)
Comment (Mprogrammer2020) [reply src/components/Chats/Chat/ChatInput.js:117]: ok sure

File: src/components/Chats/Chat/ChatInput.js (hunk 7 above)
Comment (Mprogrammer2020) [reply src/components/Chats/Chat/ChatInput.js:127]: ok sure

File: src/components/Chats/Chat/ChatInput.js (hunk 6 above)
Comment (Mprogrammer2020) [reply src/components/Chats/Chat/ChatInput.js:113]: Ok I will try

Comment (Mprogrammer2020) [reply src/components/Chats/Chat/ChatInput.js:113]: When I try with array functions it always giving me errors that's why I used for loop.

File: src/components/Chats/Chat/ChatInput.js (hunk 8 above)
Comment (Mprogrammer2020) [reply src/components/Chats/Chat/ChatInput.js:152]: To make it work on web version I have to use input and I also need to call it without actually clicking on it.

File: src/components/Chats/Chat/ChatInput.js (hunk 3 above)
Comment (Mprogrammer2020) [reply src/components/Chats/Chat/ChatInput.js:96]: Removed

File: src/components/Chats/Chat/ChatInput.js [hunk 9]
```diff
@@ -85,14 +91,52 @@ const ChatInput = (props) => {
...
+      })
+
+      let jsonRes = await response.json()
+      return jsonRes.url
+
+    });
+
+    let data = await Promise.all(uploads)
+    // Urls of images
+    console.log(data)
```
Comment (jeznag) [inline src/components/Chats/Chat/ChatInput.js:122]: you need to store these URLs so you can send them to the backend

File: src/components/Chats/Chat/ChatInput.js (hunk 8 above)
Comment (jeznag) [reply src/components/Chats/Chat/ChatInput.js:135]: have you tested on Android to make sure it works there?

File: src/pages/CallsPage.tsx [hunk 10]
```diff
@@ -1,27 +1,125 @@
...
       <IonContent fullscreen>
         <IonHeader collapse="condense">
           <IonToolbar>
             <IonTitle size="large">Calls</IonTitle>
           </IonToolbar>
         </IonHeader>
-        <div>Calls Page</div>
       </IonContent>
+
+      <IonFooter>
```
Comment (jeznag) [inline src/pages/CallsPage.tsx:59]: I don't think IonFooter is the correct component to use.

Also we will be showing calls here later. The keypad should go in its own component.

There should be a button on the calls page to show the keypad.

File: src/components/Tabs.js [hunk 11]
```diff
@@ -36,11 +36,9 @@ const Tabs = (props) => {
         <IonTabButton tab="tab2" href="/Dashboard/contact">
           <IonIcon icon={peopleOutline} />
         </IonTabButton>
-        {/*
         <IonTabButton tab="tab3" href="/Dashboard/call">
-          <IonIcon icon={callOutline} />
+          <IonIcon icon={keypadOutline} />
```
Comment (jeznag) [inline src/components/Tabs.js:40]: keep as callOutline. Also keep this commented out for now. (You can uncomment it for testing but I don't want to show it to end users yet)

File: public/index.html [hunk 12]
```diff
@@ -25,26 +33,24 @@
...
-      gtag('config', 'G-9N0SZPBFW1');
-    </script>
-  </head>
-  <body>
-    <noscript>You need to enable JavaScript to run this app.</noscript>
-    <div id="root" />
-    <!--
+  <title>Twilio SMS/Call for Zoho CRM</title>
+  <!-- Global site tag (gtag.js) - Google Analytics -->
+  <!-- <script async src="https://www.googletagmanager.com/gtag/js?id=G-9N0SZPBFW1"></script>
```
Comment (jeznag) [inline public/index.html:38]: oh we already had it there. You can keep the old way.

However, I've checked GA and it's not working properly. It's only showing a single page currently.

Can you investigate how to make it log page views when they switch from the inbox to the contact view?

I'd also like to log events when they:
1. send an SMS
2. pull to refresh
3. do a search

File: src/components/Chats/Chat/ChatInput.js (hunk 9 above)
Comment (Mprogrammer2020) [reply src/components/Chats/Chat/ChatInput.js:122]: Yes but we haven't the backend ready yet for this.

File: src/components/Tabs.js (hunk 11 above)
Comment (Mprogrammer2020) [reply src/components/Tabs.js:40]: ok

File: src/components/Chats/Chat/ChatInput.js (hunk 8 above)
Comment (Mprogrammer2020) [reply src/components/Chats/Chat/ChatInput.js:135]: Yes it is working fine but for now it is picking the existing images from gallery. I will work on its camera feature also.

File: public/index.html (hunk 12 above)
Comment (Mprogrammer2020) [reply public/index.html:38]: Ok

File: src/pages/CallsPage.tsx (hunk 10 above)
Comment (Mprogrammer2020) [reply src/pages/CallsPage.tsx:59]: So should I remove the footer then it will go up or you want anything else to keep this at bottom?

Comment (jeznag) [reply src/pages/CallsPage.tsx:59]: the dial pad should only display if you click on a button

File: src/components/Chats/Chat/ChatInput.js [hunk 13]
```diff
@@ -151,9 +189,30 @@ const ChatInput = (props) => {
...
+          setActionsPopover({ showPopover: false, event: undefined })
+        }>
+        <IonList>
+          <IonItem onClick={() => {
+            showImagePicker(true)
+          }}>
+            <IonLabel>Attachment</IonLabel>
+            <IonIcon size="large" icon={attach} />
+          </IonItem>
+          <IonItem onClick={(e) => { setShowQuickRepliesPopover({ showQuickRepliesPopover: true, event: e }); }}>
```
Comment (jeznag) [inline src/components/Chats/Chat/ChatInput.js:209]: hide this for now until the backend is ready

Comment (Mprogrammer2020) [reply src/components/Chats/Chat/ChatInput.js:212]: done

File: src/components/Chats/Chat/ChatInput.js [hunk 14]
```diff
@@ -85,13 +99,57 @@ const ChatInput = (props) => {
...
       style={{
-        padding: '10px 20px',
+        padding: '10px 0px',
         '--border-radius': '20px',
         '--background': '#f5f5f5'
       }}>
+      <input capture onChange={handleUploadImages} style={{ display: 'none' }} ref={cameraRef} accept={'image/*'} type="file" />
+      <input multiple onChange={handleUploadImages} style={{ display: 'none' }} ref={galleryRef} accept={'image/*'} type="file" />
+
+      {/* <IonButton
```
Comment (jeznag) [inline src/components/Chats/Chat/ChatInput.js:144]: sending MMS is working now? If so we can show this (but don't show the template button)

File: src/components/Chats/Chat/ChatInput.js (hunk 13 above)
Comment (jeznag) [reply src/components/Chats/Chat/ChatInput.js:212]: Attachments should be shown but not templates

File: src/environments/environment.js [hunk 15]
```diff
@@ -1,15 +1,16 @@
 // Dev
-//const API_BASE_END_POINT = 'https://delugeonaluge.com/mobile-app/local/public';
-//const MOBILE_URL = 'http://10.0.2.2:3000/Login'
+// const API_BASE_END_POINT = 'https://delugeonaluge.com/mobile-app/local/public';
```
Comment (jeznag) [inline src/environments/environment.js:2]: this can be moved to a .env file

File: src/components/Chats/Chat/ChatInput.js (hunk 14 above)
Comment (Mprogrammer2020) [reply src/components/Chats/Chat/ChatInput.js:144]: done

File: src/components/Chats/Chat/ChatInput.js (hunk 13 above)
Comment (Mprogrammer2020) [reply src/components/Chats/Chat/ChatInput.js:212]: done

File: src/components/Chats/Chat/ChatHeader.js [hunk 16]
```diff
@@ -40,10 +66,44 @@ const ChatHeader = (props) => {
...
-        <IonButtons slot="end">
-          <IonButton onClick={onClose}>Close</IonButton>
+        <IonButtons slot="end" >
+
+          <IonButton onClick={onClose} style={{ margin: 0 }}>
+            <IonIcon style={{ height: 30, width: 30, }} icon={checkmark} />
+            <IonLabel style={{ fontSize: 8 }}>No Reply<br />Needed</IonLabel>
+          </IonButton>
+
+          <IonButton
```
Comment (jeznag) [inline src/components/Chats/Chat/ChatHeader.js:76]: can you hide this button for now until we have call functionality done?

File: src/components/Chats/Chat/ChatInput.js [hunk 17]
```diff
@@ -47,29 +60,83 @@ const ChatInput = (props) => {
...
+        mms_content: JSON.stringify(showImages),
+        send_from: fromNumber,
+        send_to: toNumber || contact.number,
+        isSending: true
+      }
+    ]);
+    setSelectedImages([])
+
+    let data = await Promise.all(uploads)
+    sendMessage(data)
```
Comment (jeznag) [inline src/components/Chats/Chat/ChatInput.js:111]: It should not send the message immediately. It should only send once you press the send button.

File: src/environments/environment.js [hunk 18]
```diff
@@ -1,15 +1,16 @@
...
+// const MOBILE_URL = 'http://localhost:3000/Login'
```
Comment (jeznag) [inline src/environments/environment.js:3]: Move the commented out code to a .env.sample

File: src/components/Chats/Chat/ChatHeader.js (hunk 16 above)
Comment (Mprogrammer2020) [reply src/components/Chats/Chat/ChatHeader.js]: done

File: src/components/Chats/Chat/ChatInput.js (hunk 17 above)
Comment (Mprogrammer2020) [reply src/components/Chats/Chat/ChatInput.js:111]: Yes it is sending only when user press send button after selection.

File: src/environments/environment.js (hunk 18 above)
Comment (Mprogrammer2020) [reply src/environments/environment.js:3]: ok

**Pull request merged**